internal/wrphandlers/qos/priority_queue.go

```diff
@@ -0,0 +1,106 @@
+// SPDX-FileCopyrightText: 2024 Comcast Cable Communications Management, LLC
+// SPDX-License-Identifier: Apache-2.0
+
+package qos
+
+import (
+	"container/heap"
+	"sync"
+
+	"github.com/xmidt-org/wrp-go/v3"
+)
+
+// PriorityQueue implements heap.Interface and holds wrp Message, using wrp.QOSValue as its priority.
+// https://xmidt.io/docs/wrp/basics/#qos-description-qos
+type PriorityQueue struct {
+	queue []wrp.Message
+	// maxQueueSize is the allowable max size of the queue based on the sum of all queued wrp message's payloads
+	maxQueueSize int
+
+	m sync.Mutex
+}
+
+// Dequeue returns the next highest priority message.
+func (pq *PriorityQueue) Dequeue() (wrp.Message, bool) {
+	defer pq.m.Unlock()
+	pq.m.Lock()
+	if pq.Len() == 0 {
+		return wrp.Message{}, false
+	}
+
+	msg, ok := heap.Pop(pq).(wrp.Message)
+
+	return msg, ok
+
+}
+
+// Enqueue queues the given message.
+func (pq *PriorityQueue) Enqueue(msg wrp.Message) {
+	defer pq.m.Unlock()
```

Comment:
@schmidtw here's the update to how we trim the queue, tested with cd and it looks like it's working as expected now